Convert `tsh play` to a streaming API

zmb3 opened the pull request:
Instead of downloading the entire session before starting playback,
tsh now streams the recording directly to the terminal. This results
in a significant reduction in "time to first byte" when playing back
large session recordings.

This has the side effect of making playback of SSH consistent with
that of desktop sessions (we no longer "skip" dead time for SSH
sessions and instead play them back at true speed). Additionally,
a new CLI flag was added to support configurable playback speeds.

Another behavior that I changed here was to make the arrow keys advance
(or rewind) by 10 seconds during playback. Prior to this, they would
move by a single "frame" (effectively a single keypress).

Fixes https://github.com/gravitational/teleport/issues/10579
Closes https://github.com/gravitational/teleport/issues/11385
Updates https://github.com/gravitational/teleport-private/issues/1024

zmb3 committed:
Convert tsh play to use the new streaming API

Instead of downloading the entire session before starting playback,
tsh now streams the recording directly to the terminal. This results
in a significant reduction in "time to first byte" when playing back
large session recordings.

This has the side effect of making playback of SSH consistent with
that of desktop sessions (we no longer "skip" dead time for SSH
sessions and instead play them back at true speed). Additionally,
a new CLI flag was added to support configurable playback speeds.

Fixes #10579
Closes #11385
Updates gravitational/teleport-private#1024

zmb3 committed:
Start to phase out the GetSessionEvents API

- Mark the old HTTP client implementations as not implemented.
  Nothing is calling these, but they remain for now until we can
  fully remove them from the ClientI interface.
- Move `tsh play --format=json` over to a streaming approach.
  This ensures that we don't load the entire session into memory
  before converting it to JSON/YAML format.

ptgott reviewed (approved):
Approving docs changes

ravicious commented on lib/client/api.go:
When using a pager and moving the playback position around the time when the pager was started, I see some weird artifacts again. You can see it throughout the video, but for example at 27s both the pager and my prompt are visible at the same time.

I'm not sure how complicated it is to handle pagers correctly though.

https://github.com/gravitational/teleport/assets/27113/4657462b-84ad-4413-962c-63dca2ba4aa8

zmb3 replied:
I'll have a closer look at this.

zmb3 replied:
Looks like the issues with a pager were always present (though more likely when we're streaming over the network) and are due to the way we "timestamp" each frame (we attempt to save the cursor position, move the cursor to the top right row, write the timestamp of the last event, then restore the cursor position).

I never liked this, as it's not clear to the user that this output is something we're adding and not part of the recording.

I think I found a fix that solves both problems - write the timestamp to the terminal title rather than inside the PTY. Seems to work on Terminal.app and iTerm2. It doesn't break anything in Teleport Connect, but it doesn't change the title of the tab either. Looks like we'd need to implement the [`onTitleChange`](https://xtermjs.org/docs/api/terminal/classes/terminal/#ontitlechange) callback.

<img width="1321" alt="image" src="https://github.com/gravitational/teleport/assets/7527103/563471a8-d616-484a-9ca3-23583bc9e571">


WDYT?

ravicious replied:
It's much better now, thanks.

Regarding `onTitleChange`, Connect updates the title to reflect CWD but that's the only thing we take into account. It's been like this since the first version of Connect. Maybe we've been doing it wrong all along and we should use `onTitleChange` instead and depend on the shell to provide the title. I'll check it and create an issue for this.

ravicious replied:
I think the reason we don't use `onTitleChange` is that on my setup (zsh) it returns something like `user@hostname:<cwd>` or `<full command with args>` when running a command. Our current title for local terminals is `<cwd> | <currently selected cluster>`.

It seems that iTerm and Terminal do some extra stuff in order to get just the cwd or just the command name. If we were to support `onTitleChange`, we'd need to figure this out. But it's not as easy as adding `terminal.onTitleChange(updateTabTitle)`.

ravicious replied:
Created the issue for updating the terminal title in Connect at #36223.

ravicious reviewed (commented):
I haven't looked that closely into the code itself, but I did play with the player a little bit.

I found some weird artifacts but they affect only pagers and scrolling to the start of the playback. Overall they're not that big of a deal and the new implementation is probably still an improvement over the existing one.

ravicious commented on docs/pages/connect-your-client/tsh.mdx:
BTW, doesn't this PR close #10578 as well?

zmb3 replied:
No, this PR only touches tsh play. I will do the web UI in a separate PR.

ravicious replied:
Ah yeah, I forgot that the Web UI does not work like Connect. 😶‍🌫️

zmb3 commented on lib/client/api.go:
```diff
@@ -1995,65 +1996,138 @@ func (tc *TeleportClient) Play(ctx context.Context, namespace, sessionID string)
+				playing = !playing
+			case keyLeft, keyDown:
+				current := time.Duration(player.LastPlayed() * int64(time.Millisecond))
+				term.Stdout().Write([]byte("\x1bc"))
```
Because the state of the screen at any point in time is a function of everything that happened up until that point.

Going backwards to time t actually means restart from time 0 and then quickly dump all events up to t. 

This is not necessary when going forwards.

zmb3 committed:
Clean up terminal escape codes

- Move more utilities into lib/client.Terminal
- Stop timestamping the frame by moving the cursor and writing
  to the terminal - this is racy. Write the timestamp to the
  terminal title instead.

Tener commented on tool/tsh/common/tsh.go:
```diff
@@ -899,6 +900,7 @@ func Run(ctx context.Context, args []string, opts ...CliOption) error {
 	// play
 	play := app.Command("play", "Replay the recorded session (SSH, Kubernetes, App, DB).")
 	play.Flag("cluster", clusterHelp).Short('c').StringVar(&cf.SiteName)
+	play.Flag("speed", "Playback speed, applicable when streaming SSH or Kubernetes sessions.").Default("1x").EnumVar(&cf.PlaySpeed, "0.5x", "1x", "2x", "4x", "8x")
```
Can we simply use float instead of string with a map converted to floats?

zmb3 replied:
We could, but I don't really want to allow a speed of 1.742236.

Tener replied:
`3.14159` is such a nice speed though 😉 I think there are sensible values which are disallowed now, and just using floats would allow them as well as make the code slightly simpler.

Not a big deal in any way, though.

Tener commented on lib/client/api.go:
```diff
@@ -2001,140 +2002,142 @@ func (tc *TeleportClient) Play(ctx context.Context, namespace, sessionID string)
+	return nil
+}
+
+func setTermSize(w io.Writer, size string) error {
```
Do we need this function? It gets called on `term`, which already has an equivalent method.

zmb3 replied:
`term.Resize` takes raw integers for width height, this takes a "term size spec" as exists in session recording events (`w:h`).

We could update this to take a term, convert the `w:h` string to two integers, and then call `term.Resize`, which would be a little less duplication but would end up being more code than we have now. Thoughts?

Tener replied:
Ah, probably not worth it then.

zmb3 committed:
Update changelog

zmb3 committed:
Move terminal functions from Unix to common build

This is valid because we initialize terminals on Windows with
ENABLE_VIRTUAL_TERMINAL_PROCESSING, so the same escape sequences
that work on Unix will also work on Windows.

This matches Microsoft's recommendation to use virtual terminal
sequences over the Windows Console API [1].

For more info, see: https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences

[1] https://learn.microsoft.com/en-us/windows/console/classic-vs-vt
